mltensor-explainer.md

```diff
@@ -0,0 +1,423 @@
+
+// ...Let's hope its memory is garbage-collected soon?
+
+// Construct a new graph which itself needs a lot of resources.
+// If the system is under memory pressure, this may fail.
+const builder = new MLGraphBuilder(mlContext);
+const constant = builder.constant(descriptor, veryLargeBufferOfWeights);
+```
+
+An `MLTensor`, `MLGraph`, and `MLContext` all have a respective `destroy()` method. Once these objects are no longer needed, the website may request that the memory associated with these resources be released... possibly so that it can run more models!
+
+```js
+// Proposed approach to resource management, with explicit destroy methods
+
+mlContext.dispatch(graph1, inputs, outputs);
+
+// We're done with `graph1`, which may contain multiple gigabytes of weights.
+// Explicitly ask for its resources to be released!
+graph1.destroy();
+
+// We can selectively release only the resources we expect won't be needed.
+destroyBuffers(inputs);
+// Don't destroy the output tensors yet, in case we want to reuse them later.
+
+// Construct a new graph which itself needs a lot of resources. Memory pressure
+// has been relieved and building this graph is much more likely to succeed. 
+const builder = new MLGraphBuilder(mlContext);
+const constant = builder.constant(descriptor, veryLargeBufferOfWeights);
+```
+
+### WebGPU Interop
+
+A privacy-conscious user wants to perform real-time selfie segmentation of a video feed on their local device.
+
+Currently, using WebNN for this task would require - for each frame - an expensive readback of `GPUBuffer` data to script, uploading the data to the ML context device (which may be the same GPU!), copying the result back to script, and then uploading the frame to be rendered back into a `GPUBuffer`. This is unlikely to be performed in real-time.
+
+An `MLTensor` may be imported into WebGPU, which in the best case provides zero-copy buffer sharing between the two APIs, and in all cases provides a synchronization mechanism between the respective WebNN and WebGPU [timelines](https://www.w3.org/TR/webgpu/#programming-model-timelines), avoiding the need for expensive synchronization via script.
+
+```js
+// Create a couple MLTensors to be used to facilitate WebGPU interop.
+const mlTensor1 = await mlContext.createTensor({..., usage: MLTensorUsage.WEBGPU_INTEROP});
+const mlTensor2 = await mlContext.createTensor({..., usage: MLTensorUsage.WEBGPU_INTEROP});
+
+const applyEffectToFrame = async () => {
+  const gpuVideoTexture = gpuDevice.importExternalTexture({source: video});
+
+  // Rent out the MLTensor to WebGPU.  
+  const tensorizedGpuBuffer = gpuDevice.importExternalBuffer(mlTensor1);
+
+  // Create a bind group for `gpuVideoTexture`, create a command encoder, etc.
+  // to "tensorize" `gpuVideoTexture` and store the result in `tensorizedGpuBuffer`
+  // ...
+
+  gpuDevice.queue.submit([tensorizationCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN.
+  tensorizedGpuBuffer.destroy();
```

Comment:
Note that due to complicated GC and SharedImage / SharedBuffer cross process lifetime handling, we need one of the two APIs to be the true owner of the underlying OS buffer. The current set up where WebGPU's .destroy() returns ownership to WebML works but that's something to be careful about.